Iodevices example

hwillemen opened the pull request:
A related but more advanced diagnostics script.

hwillemen committed:
pybricks.iodevices.PUPDevice: Add more devices.

Added:
- non-uart DCMotors  (Medium and Train)
- non-uart Light
- SPIKE 3x3 Color Matrix
- SPIKE Small Angular Motor

laurensvalk committed:
pybricks.iodevices.PUPDevice: Add device class as comments.

hwillemen committed:
Diagnostics script example

see pupdevice.rst for description.
A lot of dynamic features for more advanced programming

hwillemen committed:
changes Laurens

pulling in changes from port_info.py

hwillemen commented:
I will look at the failed linter tomorrow.

laurensvalk commented:
Thanks for the submission. For larger scripts like these, it may be better to submit it to the pybricks-projects repository. Feel free to post it as an issue there first, so we can help you find a good location for it.

hwillemen committed:
examples.iodevices diagnostics lint #1

Fixed the first batch of lint errors

hwillemen commented:
My thinking was that this script is a more extended version of the iodevices example.
Since it is not related to an actual project I thought the projects repo would be less suitable.
If you do not agree with that and still thinks it should go under projects let me know.
But let me use/abuse this branch to fix a the linting errors. so a few more commits coming.

from pybricks.pupdevices import * is not allowed?
Would "import pybricks.pupdevices as pupdev" be better?
I do not want to type a very long list there.

hwillemen committed:
examples.iodevices diagnostics lint #2

2nd batch of lint errors

hwillemen committed:
satisfy linter, cleanup and small bugs

changed the "from pybricks.pupdevices import *" to the full list.
put port list creation in GetPorts().
cleaned up ConnectToDevice() and collapsed similar code.
fixed some remaining linter errors.
fixed small bugs regarding DIAGNOSTICS_PERIOD and DCMotor.
added IMU to the diagnostics.

satisfy linter

hwillemen committed:
doc8 linter errors

linter

hwillemen commented:
The linting errors are fixed now and I have taken another look at pybricks-projects.
Since this is a very generic diagnostics script, it does not seem to fit in any of those categories;
it is not a project related to any set.
Also not really a tutorial.
And I think it also does not fit in the remix category.

Since it is built upon the iodevices module I think it fits best in the api.

I do agree it is a large script though, but I am not sure if scrolling is a problem for users.

If it is, a scroll box with copy button as for projects scripts:
https://pybricks.com/projects/sets/city/60198-cargo-train/slope-detection/#program
might be an option.
I do not know how to do that in rst though.
